fontforge/utanvec.h

```diff
@@ -36,12 +36,15 @@
 #define BP_REV(v) (BasePoint) { -(v).x, -(v).y }
```

Comment:
I'd be thrilled to switch to inline functions. I didn't know they were "allowed". (Are you *sure* they're "allowed"? I haven't run into any in other sources and some people don't like/trust them.)

Reply:
Oops, I guess I was thinking more about how it worked under C++.

I personally don't have anything against them, and I think it reads easier, while also preventing multiple evaluations of parameters, unlike the macros.

Might need to stick a `static` in front of the inline declarations I suggested, to prevent linkage issues ([[1](https://stackoverflow.com/questions/34139737/inline-function-in-header-file-in-c)], [[2](https://www.greenend.org.uk/rjk/tech/inline.html)]).

Reply:
Yeah, it does seem like static inline is the way to go for these, if we can. 

As I indicated I am very happy to use C99 static inline functions in preference to macros. @frank-trampe would that be OK? Can we add that as an optional FontForge semantic?

Reply:
I already thought it was allowed because we previously agreed all C99 features are allowed, so I support

Reply:
Just clarifying my stance a little, I'm ok with inline functions in headers to be used judiciously; i.e. where it makes sense / where the function is trivially inlinable. I wouldn't be ok with trying to do this with more complex functions. 

So basically use common sense ;)

Reply:
Of course, no code is ever guaranteed to be accepted, and just because something is allowed doesn't mean it will be accepted :-)

Reply:
I primarily have a "when in Rome" approach to projects. I certainly don't indent my own code the way `spinestroke.c` is indented. 

The topic of discussion here, as I see it, is not as much how to avoid function-call overhead (which is relatively small these days) but how to symbolically represent a or chain of computations that you still want to be locally optimized.

GW's approach, it seems, was to do it in his head and write everything out every time. That's ... not ideal practice. I've been using macros because those were the standard approach at the time people were writing GW-like code. As @jtanx implies, the general problem with macros is that the compiler loses the syntax tree and the optimizer can't always put it back together. 

Inline functions -- and in C99 in particular *static* inline functions -- are the better way of handling the problem. The compiler gets the tree and therefore can locally optimize everything with much better information. With the "static" it has the option of bailing out by adding a local function. (Naively one might think that if it's non-inlined once it will never be inlined, but that actually depends on whether there are any constant/compile-time-determinable arguments.)